wfe2/wfe.go

```diff
@@ -530,13 +534,256 @@ func (wfe *WebFrontEndImpl) regHoldsAuthorizations(ctx context.Context, regID in
 	return !missingNames, nil
 }
 
-// RevokeCertificate is used by clients to request the revocation of a cert.
-func (wfe *WebFrontEndImpl) RevokeCertificate(ctx context.Context, logEvent *requestEvent, response http.ResponseWriter, request *http.Request) {
-	// RevokeCertificate is a NOP for WFEv2 at the present time. It needs unique JWS
-	// validation compared to other ACME v2 endpoints.
-	wfe.sendError(response, logEvent,
-		probs.ServerInternal("RevokeCertificate is not presently implemented for ACME v2"), nil)
-	return
+// authenticateRevocation is a callback function that can be used to validate
+// that the outer JWS of a revocation request is properly authenticated. Since
+// there are two ways a revocation request can be authenticated it is the
+// responsibility of the implementation to extract the outer JWS payload (and if
+// applicable, an associated account ID). If the request is not properly
+// authenticated a problem should be returned. Note that the account ID returned
+// _may_ be 0 if the request is self-authenticated and not associated with an
+// account.
+type authenticateRevocation func(*jose.JSONWebSignature, *http.Request, context.Context, *requestEvent) ([]byte, int64, *probs.ProblemDetails)
+
+// authorizedToRevokeCert is a callback function that can be used to validate if
+// a given requester is authorized to revoke the certificate parsed out of the
+// revocation request from the inner JWS. If the requester is not authorized to
+// revoke the certificate a problem is returned.
+type authorizedToRevokeCert func(*x509.Certificate) *probs.ProblemDetails
+
+// processRevocation accepts the outer JWS for a revocation request along with
+// two callback functions used to authenticate the JWS and decide if the
+// requester is authorized to revoke a given certificate. If the request can not
+// be authenticated or the requesteris not authorized to revoke the certificate
+// requested a problem is returned. Otherwise the certificate is marked revoked
+// through the SA.
+func (wfe *WebFrontEndImpl) processRevocation(
+	outerJWS *jose.JSONWebSignature,
+	authenticateJWS authenticateRevocation,
+	authorizedToRevoke authorizedToRevokeCert,
+	request *http.Request,
+	context context.Context,
+	logEvent *requestEvent) *probs.ProblemDetails {
+	// Authenticate the outer JWS for this revocation request and retreive the JWS
+	// body bytes and account ID (if applicable)
+	jwsBody, acctID, prob := authenticateJWS(outerJWS, request, context, logEvent)
```

Comment:
Since `authenticateJWS` is always the first thing called, wouldn't it be simpler to do the work directly in `revokeCertByKeyID` and `revokeCertByJWK` and pass in `jwsBody, acctID`?